lib/models/StreamSplitter.js

```diff
@@ -0,0 +1,131 @@
+    }
+
+    _write(chunk, encoding, callback) {
+        Promise.resolve()
+            .then(() => {
+                // In order to start writing a chunk, we must first create
+                // a file system reference for it
+                if (this.fileDescriptor === null) {
+                    return this._newChunk();
+                }
+                return undefined;
+            })
+            .then(() => {
+                const overflow = this.currentChunkSize + chunk.length - this.maxChunkSize;
+
+                // If the chunk is bigger than the defined max chunk size,
+                // we need two passes to process the chunk
+                if (overflow > 0) {
+                    return this._writeChunk(chunk.slice(0, chunk.length - overflow))
+                        .then(this._finishChunk.bind(this))
+                        .then(this._newChunk.bind(this))
+                        .then(() => {
+                            return this._writeChunk(
+                                chunk.slice(chunk.length - overflow, chunk.length)
+                            );
+                        })
+                        .then(() => callback())
+                        .catch(callback);
+                }
+
+                // The chunk fits in the max chunk size
+                return this._writeChunk(chunk)
+                    .then(() => callback())
+                    .catch(callback);
+            })
+            .catch(callback);
+    }
+
+    _final(callback) {
+        if (this.fileDescriptor === null) {
+            callback();
+        }
+        else {
+            this._finishChunk()
+                .then(() => callback())
+                .catch(callback);
+        }
+    }
+
+    _writeChunk(chunk) {
+        return new Promise((resolve, reject) => {
+            fs.write(this.fileDescriptor, chunk, (err) => {
+                if (err) {
+                    return reject(err);
+                }
+
+                this.currentChunkSize += chunk.length;
+                return resolve();
+            });
+        });
+    }
+
+    _finishChunk() {
+        if (this.fileDescriptor === null) {
+            return Promise.resolve();
+        }
+
+        return new Promise((resolve, reject) => {
+            fs.close(this.fileDescriptor, (err) => {
+                if (err) {
+                    return reject(err);
+                }
+
+                this.emit('chunkFinished', this.currentChunkPath);
+
+                this.currentChunkPath = null;
+                this.fileDescriptor = null;
+                this.currentChunkSize = null;
+
+                this.part += 1;
+
+                return resolve();
+            });
+        });
+    }
+
+    _newChunk() {
+        return new Promise((resolve, reject) => {
+            this.currentChunkPath = path.join(this.directory, `${this.filenameTemplate}-${this.part}`);
+            fs.open(this.currentChunkPath, 'w', (err, fd) => {
```

Comment:
I am wondering why you are using file descriptors here directly, instead of using streams? Is there some benefit? Easier usage?

Reply:
We do use streams, but we create a tmp file for the stream to write to.

Reply:
Yes, but we have you decided to use `fs.open` and `fs.write(fd, ...)` directly instead of using `fs.createWriteStream` and `steam.write`? Are these too high level?

Reply:
Right, the pattern of `open` and `write` has been used more often in this project. Both would do the job. I suppose we use it here to do some separation of concerns. `createWriteStream` takes care of opening and closing the stream as well, but we do some side-effects when that happens (so when we call `open` for instance).